1 add pagination

AlexNguyenJJ opened the pull request:
Implemented LimitOffset pagination
 - Created Custom Pagination class `MoviesPagination` for more control over settings
 - Added validation functions for limit and offset params
 - `MoviesPagination` returns formatted links to next and previous pages in pagination in accordance to correct hostname
 - Provided more commenting and documentation to functions.

AlexNguyenJJ committed:
Add pagination functionality

Created custom pagination class for more control
Implemented simple pagination using limit offset
Added some more detail to function usage for movie_list() in view.py
TODO: Remove all hosts except for the production site https://winearth.sdsc.edu from the CORS_ORIGIN_WHITELIST in settings.py

AlexNguyenJJ committed:
FIX: Decode URLs and add dev setting for hostname in paginated responses

Leveraged urllib package to decode URLs that encode special characters like '/' to be compatible with existing functions
Added validation functions for limit and offset params
Added to customized pagination class custom links to next and previous to match hostname

AlexNguyenJJ committed:
Reset CORS_ORIGIN_WHITELIST to production settings

AlexNguyenJJ committed:
Fix URL decoding

Fix so that decoding only takes place when there is something to decode. Previous version tried to decode on NoneType which is not allowed

kevincoakley commented on api/views.py:
```diff
@@ -76,28 +88,62 @@ def index(request):
 
+        # Decode the URL
+        decoded_request = urllib.parse.unquote(request.build_absolute_uri())
+        print(decoded_request)
```
Don't include print statements

kevincoakley commented on api/mypaginations.py:
```diff
@@ -0,0 +1,20 @@
+class MoviesPagination(LimitOffsetPagination):
+  default_limit = 100
+  max_limit = 20
+  DEV_MODE = False # Change to true when running standalone docker container for winearth-api
```
Not clear what the DEV_MODE and host variables are used for. If they are not needed don't include them.

kevincoakley commented on api/mypaginations.py:
```diff
@@ -0,0 +1,20 @@
+
+  def get_paginated_response(self, data):
+    nl = self.get_next_link()
+    print(f'next link: {nl}')
```
Don't include print statements

kevincoakley commented on api/mypaginations.py:
```diff
@@ -0,0 +1,20 @@
+    nl = self.get_next_link()
+    print(f'next link: {nl}')
+    pl = self.get_previous_link()
+    print(f'prev link: {pl}')
```
Don't include print statements

kevincoakley commented on api/mypaginations.py:
```diff
@@ -0,0 +1,20 @@
+  DEV_MODE = False # Change to true when running standalone docker container for winearth-api
+  host = 'http://127.0.0.1' if DEV_MODE else 'https://winearth.sdsc.edu'
+
+  def get_paginated_response(self, data):
```
Please comment on what this code does. I have not seen other implementations online do this and it isn't clear why the response would include http://127.0.0.1

AlexNguyenJJ replied:
My mistake, I didn't implement it correctly in the body of the code. It's supposed to serve as an easy way for me to switch between development testing and what is supposed to be on production. When in development with the docker compose file, the dev site is on 127.0.0.1 and so in order to test if the links work, I have to set the host manually. When in production however, it will use the hostname of the production site `https://winearth.sdsc.edu`. I'm going to fix this so it works as intended. If you don't like my approach to this and would instead rather have me swap it out manually during development i.e. not having the `DEV_MODE` and `host` variables and changing the hostname string directly from `https://winearth.sdsc.edu` to `http://127.0.0.1`, then I can do that as well.

kevincoakley commented on api/views.py:
```diff
@@ -25,6 +27,16 @@ def validate_length(length_str):
     length_regex = r"^\d+$"
     if not re.match(length_regex, length_str):
         raise ValidationError(f"length {length_str} is not in the correct format.")
+    
```
It looks like the functions validate_length, validate_limit and validate_offset all do the same thing. The code would be much simpler if there was only one function to validate integers.

AlexNguyenJJ replied:
Will do

kevincoakley reviewed (changes requested):
Please read the following comments. 

Also, we should use the same coding style. Use black (https://black.readthedocs.io/en/stable/index.html) to format your code and check you code with flake8 using:
`flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics`
`flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics`

I added black format checking and the above flake8 lint checks to the github action, so only code that passes those test will be accepted.

AlexNguyenJJ committed:
Address PR requested changes

 - Consolidated validation function for integers
 - Corrected DEV_MODE implementation in mypaginations.py

AlexNguyenJJ committed:
Linting changes

AlexNguyenJJ committed:
Merge branch 'main' into 1-add-pagination

AlexNguyenJJ committed:
More linting to pass checks

AlexNguyenJJ committed:
Linting changes that were missed

kevincoakley commented:
Superseded by https://github.com/windows-on-earth/winearth-api/pull/21
